samples/template/src/main.c

```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <sid_api.h>
+#include <sid_hal_reset_ifc.h>
+#include <pal_init.h>
+#include <app_ble_config.h>
+#include <app_subGHz_config.h>
+#include <dk_buttons_and_leds.h>
+#include <zephyr/kernel.h>
+#include <zephyr/smf.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(app, CONFIG_LOG_DEFAULT_LEVEL);
+
+static struct k_thread app_thread;
+K_THREAD_STACK_DEFINE(app_thread_stack, CONFIG_SIDEWALK_THREAD_STACK_SIZE);
+
+typedef enum app_events {
+	APP_EVENT_SIDEWALK,
+	APP_EVENT_ERROR,
+	APP_EVENT_READY,
+	APP_EVENT_NOT_READY,
+	APP_EVENT_SEND,
+	APP_EVENT_FACTORY_RESET,
+	APP_EVENT_LINK_SWITCH
+} app_event_t;
+
+enum app_state {
+	STATE_RUNNING,
+	STATE_INIT,
+	STATE_SIDEWALK_NOT_READY,
+	STATE_SIDEWALK_CONNECTING,
+	STATE_SIDEWALK_READY,
+};
+
+struct app_sm {
+	struct smf_ctx ctx;
+	struct k_msgq msgq;
+	app_event_t event;
+};
+
+struct sid_ctx_s {
+	struct app_sm sm;
+	struct sid_handle *handle;
+	struct sid_config config;
+};
+
+struct app_msg {
+	struct sid_msg msg;
+	struct sid_msg_desc desc;
+};
+
+static int app_event_send(app_event_t event);
+
+static void state_running_run(void *o);
+static void state_init_entry(void *o);
+static void state_init_run(void *o);
+static void state_ready_run(void *o);
+static void state_not_ready_run(void *o);
+static void state_connecting_entry(void *o);
+static void state_connecting_run(void *o);
+
+static const struct smf_state app_states[] = {
```

Comment:
There are two machine states, first one handles the core functionalities like sid_process, DFU, reset.
The second one is for the application,

This is seen in the graph from doc/smf_graph.puml

It would be nice to separate those state machines into separate files

The outer SMF is common for all apps, so it may be inside main with the sidewalk callbacks
the application SMF can be placed in separate file.

Reply:
I'm not sure if zephyr state machine allows to have a parent form different state machine instance. But it's worth to check